File: 11-container-with-most-water/11-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        water = 0
        l = 0
        r = len(height) - 1
        while l < r:
            h = min(height[l], height[r])
            area = h * (r - l)
            water = max(water, area)
            while height[l] <= h and l < r:
                l += 1
            while height[r] <= h and l < r:
                r -= 1
        return water

File: 11-container-with-most-water/11-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int water = 0, l = 0, r = height.size() - 1;
        while (l < r) {
            int h = min(height[l], height[r]);
            int area = h * (r - l);
            water = max(water, area);
            while (height[l] <= h && l < r) l++;
            while (height[r] <= h && l < r) r--;
        }
        return water;
    }
};